app/src/main/java/com/github/sdp/ratemyepfl/model/items/Course.kt

```diff
@@ -20,26 +20,79 @@ data class Course(
         return "$id $title"
     }
 
-    companion object {
-        /**
-         * Converts the json data in a Course object
-         *
-         * @return a course object if it contains the correct data, null otherwise
-         */
-        fun DocumentSnapshot.toCourse(): Course? {
-            val title: String? = getString("title")
-            val section: String? = getString("section")
-            val teacher: String? = getString("teacher")
-            val credits: Int? = getString("credits")?.toInt()
-
-            return if (title != null && section != null && teacher != null && credits != null) {
-                Course(title, section, teacher, credits, id)
-            } else {
-                null
-            }
+    /**
+     * Builder to create a course step by step
+     */
+    class Builder : ReviewableBuilder<Course> {
+        private var title: String? = null
+        private var section: String? = null
+        private var teacher: String? = null
+        private var credits: Int? = null
+        private var id: String? = null
+        private var cycle: String? = null
+        private var session: String? = null
+        private var grading: String? = null
+        private var language: String? = null
+
+        fun setTitle(title: String?) = apply {
+            this.title = title
+        }
+
+        fun setSection(section: String?) = apply {
+            this.section = section
+        }
+
+        fun setTeacher(teacher: String?) = apply {
+            this.teacher = teacher
+        }
+
+        fun setCredits(credits: Int?) = apply {
+            this.credits = credits
+        }
+
+        fun setId(id: String?) = apply {
+            this.id = id
+        }
+
+        fun setCycle(cycle: String?) = apply {
+            this.cycle = cycle
+        }
+
+        fun setSession(session: String?) = apply {
+            this.session = session
+        }
+
+        fun setGrading(grading: String?) = apply {
+            this.grading = grading
+        }
+
+        fun setLanguage(language: String?) = apply {
+            this.language = language
+        }
+
+        override fun build(): Course {
+            val title = this.title
+            val section = this.section
+            val teacher = this.teacher
+            val credits = this.credits
+            val id = this.id
+            val cycle = this.cycle
+            val session = this.session
+            val grading = this.grading
+            val language = this.language
+
+            // Check for mandatory properties (i.e., non nullable)
+            return if (title != null &&
+                section != null &&
+                teacher != null &&
+                credits != null &&
+                id != null
+            ) {
```

Comment:
In my opinion, checking if all attributes are different than null is the same than using a try catch with "!!"

Reply:
I understand that using null makes it heavier, but to me, a try-catch is not meant to replace control flows. In my new commit, I propose a slightly different approach with the asMandatory function. I made it more concise, readable, and lighter than the try-catch. If you want to give me feedback, I would love to hear it :)